thorclient/wsclient/client.go

```diff
@@ -0,0 +1,172 @@
+	conn, err := c.connect("/subscriptions/block", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[blocks.JSONCollapsedBlock](conn), nil
+}
+
+// SubscribeTransfers subscribes to transfer events based on the provided query.
+// It returns a Subscription that streams transfer messages or an error if the connection fails.
+func (c *Client) SubscribeTransfers(query string) (*common.Subscription[*subscriptions.TransferMessage], error) {
+	conn, err := c.connect("/subscriptions/transfer", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[subscriptions.TransferMessage](conn), nil
+}
+
+// SubscribeTxPool subscribes to pending transaction pool updates based on the provided query.
+// It returns a Subscription that streams pending transaction messages or an error if the connection fails.
+func (c *Client) SubscribeTxPool(query string) (*common.Subscription[*subscriptions.PendingTxIDMessage], error) {
+	conn, err := c.connect("/subscriptions/txpool", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[subscriptions.PendingTxIDMessage](conn), nil
+}
+
+// SubscribeBeats2 subscribes to Beat2 messages based on the provided query.
+// It returns a Subscription that streams Beat2 messages or an error if the connection fails.
+func (c *Client) SubscribeBeats2(query string) (*common.Subscription[*subscriptions.Beat2Message], error) {
+	conn, err := c.connect("/subscriptions/beat2", query)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect - %w", err)
+	}
+
+	return subscribe[subscriptions.Beat2Message](conn), nil
+}
+
+// subscribe starts a new subscription over the given WebSocket connection.
+// It returns a read-only channel that streams events of type T.
+func subscribe[T any](conn *websocket.Conn) *common.Subscription[*T] {
+	// Create a new channel for events
+	eventChan := make(chan common.EventWrapper[*T], 1_000)
+	var closed bool
+
+	// Start a goroutine to handle receiving messages from the WebSocket connection.
+	go func() {
+		defer close(eventChan)
+		defer conn.Close()
+
+		for {
+			conn.SetReadDeadline(time.Now().Add(readTimeout))
+			var data T
+			// Read a JSON message from the WebSocket and unmarshal it into the data.
+			err := conn.ReadJSON(&data)
+			if err != nil {
+				if !closed {
+					// Send an EventWrapper with the error to the channel.
+					eventChan <- common.EventWrapper[*T]{Error: fmt.Errorf("%w: %w", common.ErrUnexpectedMsg, err)}
+				}
+				return
+			}
+
+			eventChan <- common.EventWrapper[*T]{Data: &data}
+		}
+	}()
+
+	return &common.Subscription[*T]{
+		EventChan: eventChan,
+		Unsubscribe: func() {
+			closed = true
+			conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+			conn.Close()
```

Comment:
`// todo add metrics` was removed since this is a client code, mostly not running with in thor program.